contrib/google-protobuf-cmake/protobuf_generate.cmake

```diff
@@ -157,23 +157,21 @@ function(protobuf_generate)
 
   set(_generated_srcs_all)
   foreach(_proto ${protobuf_generate_PROTOS})
-    get_filename_component(_abs_file ${_proto} ABSOLUTE)
-    get_filename_component(_abs_dir ${_abs_file} DIRECTORY)
-    get_filename_component(_basename ${_proto} NAME_WE)
-    file(RELATIVE_PATH _rel_dir ${CMAKE_CURRENT_SOURCE_DIR} ${_abs_dir})
-
-    set(_possible_rel_dir)
-    if (NOT protobuf_generate_APPEND_PATH)
-        set(_possible_rel_dir ${_rel_dir}/)
-    endif()
+    # The protobuf compiler doesn't return paths to the files it generates so we have to calculate those paths here:
+    # _abs_file - absolute path to a .proto file,
+    # _possible_rel_dir - relative path to the .proto file from some import directory specified in Protobuf_IMPORT_DIRS,
+    # _basename - filename of the .proto file (without path and without extenstion).
+    get_proto_absolute_path(_abs_file "${_proto}" ${_protobuf_include_path})
+    get_proto_relative_path(_possible_rel_dir "${_abs_file}" ${_protobuf_include_path})
+    get_filename_component(_basename "${_abs_file}" NAME_WE)
```

Comment:
Without my changes here it the cmake script returned incorrect paths for generated source files after compiling .proto files. As a result variables `prometheus_protobufs_sources` and `prometheus_protobufs_headers` [here](https://github.com/ClickHouse/ClickHouse/blob/1ef5bca59164bd4ec00743e0f6a5d7cf17c077ef/contrib/prometheus-protobufs-cmake/CMakeLists.txt#L30) were incorrect and the build failed.